### ISSUE TITLE
JavadocType: False negative for unknown tag in the format `abc.xyz`

### ISSUE THREAD

**Issue by sktpy** [checkstyle/checkstyle#14618]
([Documentation](https://checkstyle.org/checks/javadoc/javadoctype.html))

Discovered this while working on the fix for https://github.com/checkstyle/checkstyle/issues/14573

Example of such usages in real-code can be found [here](https://github.com/search?q=repo%3Aapache%2Fant%20%40ant.task&type=code).

**Test.java** :ant: 
```java
/**
 * @ant.task category="test"
 * @abc.xyz
 */
public class Test {}
```
<br>

**config.xml**
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="JavadocType"/>
  </module>
</module>
```
<br>

**Current CLI:**
```java
java -jar checkstyle-10.14.0-all.jar -c config.xml Test.java
Starting audit...
Audit done.
```
<br>

**Expected:**
```java
Starting audit...
[ERROR] Test.java:2:4: Unknown tag 'ant.task'. [JavadocType]
[ERROR] Test.java:3:4: Unknown tag 'abc.xyz'. [JavadocType]
Audit done.
Checkstyle ends with 2 errors.
```

**Comment by rnveach** [comment]
While you say this is a false negative, you don't explain why. `allowUnknownTags` default is false, and these don't look like standard tags.

**Comment by sktpy** [comment]
> these don't look like standard tags.
They aren't.

> `allowUnknownTags` default is false
That's why we should have violation.

**Comment by rnveach** [comment]
Sorry, I swapped expected and actual CLI output. Ignore my comment.

**Comment by sktpy** [comment]
No worries :)

**Comment by sktpy** [comment]
Before proceeding, how do we feel about adding a new property `allowedUnkownTags`?
It wouldn't complicate the check too much from the implementation point of view, but would be very user friendly, as this appears to be a popular case where documentation generation is based off a custom tags.

Because I feel like for those cases, our users are probably setting `allowUnknownTags` property as true,
which just puts a blanket over all the unknown tags, increasing the chances of unintended side effects.

**Comment by nrmancuso** [comment]
>Before proceeding, how do we feel about adding a new property allowedUnkownTags?

This is a logical proposal, anyone is welcome to open issues for further discussion :) 

In my limited experience, I find it is easier to make incremental changes to deliver features/bug fixes to users faster and get actual feedback from a few users. If we have some reliable suppression and this is enough, then it is enough :) So, we deliver this bug fix with a suppression that can be extended to deliver functionality like `allowedUnkownTags` would, then if users request a new property to make this easier, then we do that. 

Otherwise, we end up with checks with a lot of properties; extra properties typically increase the potential for bugs and maintenance overhead. Also, we want to avoid falling into the rabbit hole of perfecting things; nothing is ever perfect.

**Comment by sktpy** [comment]
**Note:** Click on "▶Heading" to expand its contents.

<details>
  <summary><b>Test.java</b></summary>

```java
/**
 * @ignored ok
 * @checked violation
 */
public class Test {
    /**
     * @ignored ok
     * @checked violation
     */
    class Test2 {

        /**
         * @WignoredW violation
         * @ignored ok
         * @author ok
         */
        static class Test3 {
            static class Test4 {

                /**
                 * @ignore violation
                 * @ignored ok
                 */
                static class Test5 {

                    /**
                     * @gnored violation
                     * @ignored ok
                     */
                    enum Test6 {

                    }
                }

                record Test7() {
                    /**
                     * @ignOred violation
                     * @ignored ok
                     */
                    record Test8() {

                    }
                }
            }

            /**
             * @checked violation
             * @ignored ok
             */
            interface Test9 {

            }
        }

        /**
         *
         * @IGNORED violation
         * @ignored ok
         */
        @interface Test10 {}
    }
}

```

</details>

<hr>

### Without Suppression:

<details>
  <summary><b>config.xml</b></summary>
  
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="JavadocType"/>
  </module>
</module>
```

</details>

<details>
  <summary><b>CLI</b></summary>
  
```cs
java -jar checkstyle-10.14.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] Test.java:2:4: Unknown tag 'ignored'. [JavadocType]
[ERROR] Test.java:3:4: Unknown tag 'checked'. [JavadocType]
[ERROR] Test.java:7:8: Unknown tag 'ignored'. [JavadocType]
[ERROR] Test.java:8:8: Unknown tag 'checked'. [JavadocType]
[ERROR] Test.java:13:12: Unknown tag 'WignoredW'. [JavadocType]
[ERROR] Test.java:14:12: Unknown tag 'ignored'. [JavadocType]
[ERROR] Test.java:21:20: Unknown tag 'ignore'. [JavadocType]
[ERROR] Test.java:22:20: Unknown tag 'ignored'. [JavadocType]
[ERROR] Test.java:27:24: Unknown tag 'gnored'. [JavadocType]
[ERROR] Test.java:28:24: Unknown tag 'ignored'. [JavadocType]
[ERROR] Test.java:37:24: Unknown tag 'ignOred'. [JavadocType]
[ERROR] Test.java:38:24: Unknown tag 'ignored'. [JavadocType]
[ERROR] Test.java:47:16: Unknown tag 'checked'. [JavadocType]
[ERROR] Test.java:48:16: Unknown tag 'ignored'. [JavadocType]
[ERROR] Test.java:57:12: Unknown tag 'IGNORED'. [JavadocType]
[ERROR] Test.java:58:12: Unknown tag 'ignored'. [JavadocType]
Audit done.
Checkstyle ends with 16 errors.
```

</details>

<hr>

### Suppressing violation for unknown tag `ignored`:

- Here we have suppressed unknown tag violation for the tag which is exactly named `ignored`.

<details>
  <summary><b>config.xml</b></summary>
  
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="JavadocType"/>
  </module>
  <module name="SuppressionSingleFilter">
    <property name="checks" value="JavadocType"/>
    <property name="message" value="'ignored'"/>
  </module>
</module>
```

</details>

<details>
  <summary><b>CLI</b></summary>
  
```cs
java -jar checkstyle-10.14.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] Test.java:3:4: Unknown tag 'checked'. [JavadocType]
[ERROR] Test.java:8:8: Unknown tag 'checked'. [JavadocType]
[ERROR] Test.java:13:12: Unknown tag 'WignoredW'. [JavadocType]
[ERROR] Test.java:21:20: Unknown tag 'ignore'. [JavadocType]
[ERROR] Test.java:27:24: Unknown tag 'gnored'. [JavadocType]
[ERROR] Test.java:37:24: Unknown tag 'ignOred'. [JavadocType]
[ERROR] Test.java:47:16: Unknown tag 'checked'. [JavadocType]
[ERROR] Test.java:57:12: Unknown tag 'IGNORED'. [JavadocType]
Audit done.
Checkstyle ends with 8 errors.
```

</details>

<hr>

**Conclusion:** 

- We are successfully able to suppress violation for the specified unknown tag via [SuppressionSingleFilter](https://checkstyle.org/filters/suppressionsinglefilter.html#SuppressionSingleFilter).
- Have sprinkled in some variations of the unknown tag (eg case-mismatch, content-mismatch) we want to ignore in various applicable places, to verify that it is suppressing exactly the specified tag in various places.

If a concise/extended version of this will be more preferred, let me know.

**Comment by sktpy** [comment]
**Note:** Click on "▶Heading" to expand its contents.

<details>
  <summary><b>Test.java</b></summary>

```java
/**
 * @ignored ok
 * @checked violation
 */
public class Test {
  /**
   * @ignored ok
   * @checked violation
   */
  class Test2 {

    /**
     * @WignoredW violation
     * @ignored ok
     * @author ok
     */
    static class Test3 {
      static class Test4 {

        /**
         * @ignore violation
         * @ignored ok
         */
        static class Test5 {

          /**
           * @gnored violation
           * @ignored ok
           */
          enum Test6 {

          }
        }

        record Test7() {
          /**
           * @ignOred violation
           * @ignored ok
           */
          record Test8() {

          }
        }
      }

      /**
       * @checked violation
       * @ignored ok
       */
      interface Test9 {

      }
    }

    /**
     *
     * @IGNORED violation
     * @ignored ok
     */
    @interface Test10 {}

    /**
     *
     * @param ignored
     */
    record Test11() {}
  }
}
```

</details>

<hr>

### Without Suppression:

<details>
  <summary><b>config.xml</b></summary>

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="JavadocType"/>
  </module>
</module>
```

</details>

<details>
  <summary><b>CLI</b></summary>

```cs
java -jar checkstyle-10.14.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] Test.java:2:4: Unknown tag 'ignored'. [JavadocType]
[ERROR] Test.java:3:4: Unknown tag 'checked'. [JavadocType]
[ERROR] Test.java:7:8: Unknown tag 'ignored'. [JavadocType]
[ERROR] Test.java:8:8: Unknown tag 'checked'. [JavadocType]
[ERROR] Test.java:13:12: Unknown tag 'WignoredW'. [JavadocType]
[ERROR] Test.java:14:12: Unknown tag 'ignored'. [JavadocType]
[ERROR] Test.java:21:20: Unknown tag 'ignore'. [JavadocType]
[ERROR] Test.java:22:20: Unknown tag 'ignored'. [JavadocType]
[ERROR] Test.java:27:24: Unknown tag 'gnored'. [JavadocType]
[ERROR] Test.java:28:24: Unknown tag 'ignored'. [JavadocType]
[ERROR] Test.java:37:24: Unknown tag 'ignOred'. [JavadocType]
[ERROR] Test.java:38:24: Unknown tag 'ignored'. [JavadocType]
[ERROR] Test.java:47:16: Unknown tag 'checked'. [JavadocType]
[ERROR] Test.java:48:16: Unknown tag 'ignored'. [JavadocType]
[ERROR] Test.java:57:12: Unknown tag 'IGNORED'. [JavadocType]
[ERROR] Test.java:58:12: Unknown tag 'ignored'. [JavadocType]
[ERROR] Test.java:64:12: Unused @param tag for 'ignored'. [JavadocType]
Audit done.
Checkstyle ends with 17 errors.
```

</details>

<hr>

### Suppressing violation for unknown tag `ignored`:

- Here we have suppressed unknown tag violation for the tag which is exactly named `ignored`.

<details>
  <summary><b>config.xml</b></summary>

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="JavadocType"/>
  </module>
  <module name="SuppressionSingleFilter">
    <property name="checks" value="JavadocType"/>
    <property name="message" value="Unknown tag 'ignored'"/>
  </module>
</module>
```

</details>

<details>
  <summary><b>CLI</b></summary>

```cs
java -jar checkstyle-10.14.0-all.jar -c config.xml Test.java
Starting audit...
Test.java:3:4: Unknown tag 'checked'. [JavadocType]
Test.java:8:8: Unknown tag 'checked'. [JavadocType]
Test.java:13:12: Unknown tag 'WignoredW'. [JavadocType]
Test.java:21:20: Unknown tag 'ignore'. [JavadocType]
Test.java:27:24: Unknown tag 'gnored'. [JavadocType]
Test.java:37:24: Unknown tag 'ignOred'. [JavadocType]
Test.java:47:16: Unknown tag 'checked'. [JavadocType]
Test.java:57:12: Unknown tag 'IGNORED'. [JavadocType]
Test.java:64:12: Unused @param tag for 'ignored'. [JavadocType]
Audit done.
Checkstyle ends with 9 errors.
```

</details>

<hr>

**Conclusion:**

- We are successfully able to suppress violation for the specified unknown tag via [SuppressionSingleFilter](https://checkstyle.org/filters/suppressionsinglefilter.html#SuppressionSingleFilter).
- Have sprinkled in some variations of the unknown tag (eg case-mismatch, content-mismatch) we want to ignore in various applicable places, to verify that it is suppressing exactly the specified tag in various places.
- Have placed input(at last) where the name of the unknown tag we want to suppress matches exactly with what would appear in the violation that we do not want to suppress here (missing param in this case), to show that it won't conflict with other violations of the Check reported for the exact name.
- `Unknown tag 'tagName'` only appears in the violation message for `javadoc.unknownTag` which is only used by JavadocType check, so technically we can also remove line `<property name="checks" value="JavadocType"/>` from the suppression but I haven't given it a deep thought yet.

If a concise/extended version of this will be more preferred, let me know.

**Comment by nrmancuso** [comment]
I am good to approve, @romani please take a look and add approved label if you are good.

**Comment by romani** [comment]
This Check is not AST based, if you start heavily test it you would never stop opening new issues and fixing them.

All not AST based Checks are kind of deprecated, we should reimplement them first.

We allow some updates to such Checks, that are easy and annoying users.

This issue doesn't looks like bothering anyone. It is valid issue but I don't think we need fix it before refactoring/redesign.

Redesign might also required, as this Checks too much validation, it is already proven to be error prone.

**Comment by sktpy** [comment]
> if you start heavily test it you would never stop opening new issues and fixing them

Appears to be the case. :lady_beetle:<sup>∞</sup> :repeat:
<br>

> we should reimplement them first.

:+1:, should I do that as the part of https://github.com/checkstyle/checkstyle/issues/11166 or is there a need for a new issue?

**Comment by romani** [comment]
Better to stay away of javadoc for a bit, there is no maintainers that can handle new design approvals for new module, it is mentally heavy operation. With current rate of PRs, it will be hard to do this.
Better to use your time in already approved issues for new modules.

**Comment by romani** [comment]
I updated issue, to put javadoc on a side.

**Comment by romani** [comment]
I am approving this issue, I placed "Attention note from maintainers:" in issue description.

**Comment by aclfe** [comment]
@romani if you agree, can we put "blocked" on this as the migration of JavadocType to ast at https://github.com/checkstyle/checkstyle/issues/19146 is taking place so we can address the false negatives after it's been migrated

**Comment by gianmarcoschifone** [comment]
@romani resolved thanks to AST migration.

```
gianmarco@Mac ~/Documents/Projects/testing-checkstyle % cat Test.java
/**
 * @ant.task category="test"
 * @abc.xyz
 */
public class Test {}

gianmarco@Mac ~/Documents/Projects/testing-checkstyle % cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="JavadocType"/>
  </module>
</module>

gianmarco@Mac ~/Documents/Projects/testing-checkstyle % java -jar checkstyle-13.9.0-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /Users/gianmarco/Documents/Projects/testing-checkstyle/Test.java:2:4: Unknown tag 'ant.task'. [JavadocType]
[ERROR] /Users/gianmarco/Documents/Projects/testing-checkstyle/Test.java:3:4: Unknown tag 'abc.xyz'. [JavadocType]
Audit done.
Checkstyle ends with 2 errors.
```
